aten/src/ATen/native/cpu/ForeachOpsKernels.cpp

```diff
@@ -0,0 +1,23 @@
+#include <ATen/Dispatch.h>
+#include <ATen/native/DispatchStub.h>
+#include <ATen/native/ForeachOps.h>
+
+namespace at { namespace native {
+namespace {
+
+static std::vector<Tensor> foreach_add_scalar_kernel_cpu(TensorList tensors, Scalar scalar) {
+  std::vector<Tensor> result;
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kBool, kBFloat16, kHalf, tensors[0].scalar_type(), "foreach_tensor_add_scalar_kernel_cpu", [&]() {
+    for (int i = 0; i < tensors.size(); i++) {
+      auto temp = tensors[i].add(scalar.to<scalar_t>());
```

Comment:
There are a few things that seem a bit dodgy here.

1. You've put this code in the `cpu/` folder, which means you want the compiler to build this with the AVX instruction set. But you are not doing any computational code; you're just redispatching to add (which itself will hit a vectorized implementation). So you're basically compiling this code for three times for no good reason. Don't put this kernel in the cpu/ folder! (And if you put it here because you're trying to use DispatchStub, stop using DispatchStub. Instead, in native_functions.yaml, you can just point it directly at cpu or cuda implementations.)

Reply:
2. You're doing dtype dispatch so that you can cast the `Scalar` input into a `scalar_t`. But I'm pretty sure all that happens when you call `add` is you reconstruct it back into a Scalar; i.e., `tensors[i].add(scalar)` would have worked just as well. In that case, you don't need the dispatch macro (and can, once again, save yourself from useful generating a pile of code for each dtype template specialization.)

Reply:
Not only that, this destroys proper type promotion behavior (e.g. when a floating point scalar is added to integral tensor, result should be floating point. This won't happen here, as scalar would be cast to integral type.